pennylane/transforms/batch_partial.py

```diff
@@ -0,0 +1,180 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the batch dimension transform.
+"""
+import functools
+import inspect
+
+import pennylane as qml
+
+
+def _convert_to_args(func, args, kwargs):
+    """
+    Given a function, convert the positional and
+    keyword arguments to purely positional arguments.
+    """
+    sig = inspect.signature(func).parameters
+
+    new_args = []
+    for i, param in enumerate(sig):
+        if param in kwargs:
+            # first check if the name is provided in kwargs
+            new_args.append(kwargs[param])
+        elif i < len(sig):
+            # next check if the argnum is provided
+            new_args.append(args[i])
+
+    return tuple(new_args)
+
+
+def batch_partial(qnode, all_operations=False, **partial_kwargs):
+    """
+    Create a wrapper function around the QNode with partially
+    evaluated parameters, which supports an initial batch dimension.
+
+    Args:
+        qnode (pennylane.QNode): QNode to partially evaluate
+        all_operations (bool): If ``True``, a batch dimension will be added to *all* operations
+            in the QNode, rather than just trainable QNode parameters.
+        partial_kwargs (dict): partially-evaluated parameters to pass to the QNode
+
+    Returns:
+        func: Function which accepts the same arguments as the QNode minus the
+        partially evaluated arguments provided, and behaves the same as the QNode
+        called with both the partially evaluated arguments and the extra arguments.
+        However, the first dimension of each argument of the returned function
+        will be treated as a batch dimension. The function output will also contain
+        an initial batch dimension.
+
+    **Example**
+
+    Consider the following circuit:
+
+    .. code-block:: python
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit(x, y):
+            qml.RX(x, wires=0)
+            qml.RY(y[..., 0], wires=0)
+            qml.RY(y[..., 1], wires=1)
```

Comment:
I would expect the circuit syntax without the `Ellipsis` indexing for a regular circuit.
Why is this used/required here? :)

Reply:
The latest commit removes the need to index the `y` variable.

To answer your question however, I'm not too familiar with what happens behind-the-scenes in `batch_transform`, but it seems necessary to make sure the right parameter is passed to the circuit, otherwise we get the ` Parameter [...] has incorrect batch dimension.` error.

I think it's worth noting that this is known behaviour. For example, in `templates.BasicEntanglerLayers` lines 188-189, we have
```python
for i in range(len(wires)):
    op_list.append(rotation(weights[..., layer, i], wires=wires[i : i + 1]))
```
and in `templates.StronglyEntanglingLayers` lines 208-216 we have
```python
for i in range(len(wires)):  # pylint: disable=consider-using-enumerate
    op_list.append(
        qml.Rot(
            weights[..., l, i, 0],
            weights[..., l, i, 1],
            weights[..., l, i, 2],
            wires=wires[i],
        )
    )
```
I don't know if this is documented somewhere though.

Reply:
Oh, my bad! This seems like a standard syntax, then, that I was not aware of. Thanks! :) 
I'll take a look next week whether the doc could be improved...